### PR TITLE
[FLOC-4323] Gce benchmarking

### PR DESCRIPTION
**Pull request by sarum90**
Fixes: https://clusterhq.atlassian.net/browse/FLOC-4323

This puts GCE the rest of the way into our provisioning code so you can now quickly spin up instances in parallel.

The approach was to pull out code that could be re-used from the GCE provisioner, and add an async version of wait_for_operation. The API calls to GCE are still syncronous, but we wait in parallel for the nodes to be brought up.

### CODE REVIEW

File: flocker/node/agents/gce.py [hunk 1]
```diff
@@ -230,6 +230,55 @@ def finished_operation_result():
...
+    :param reactor: The twisted ``IReactorTime`` provider to use to schedule
+        delays.
+    :param compute: The GCE compute python API object.
+    :param operation: A dict representing a pending GCE operation resource.
+        This can be either a zone or a global operation.
+    :param timeout_steps: Iterable of times in seconds to wait until timing out
+        the operation.
+
+    :returns Deferred: A Deferred firing with the concluded GCE operation
+        resource or calling its errback operation times out.
```
Comment (wallrj) [inline flocker/node/agents/gce.py]: typo "if it times out" (but not worth fixing this if you have to re-run all the builds)

File: flocker/provision/_gce.py [hunk 2]
```diff
@@ -431,44 +435,31 @@ class GCEInstanceBuilder(PClass):
...
-            self.compute, operation, timeout_steps=[5]*60)
-
-        if not operation_result:
-            raise ValueError(
-                "Timed out waiting for creation of VM: {}".format(
-                    instance_name
-                )
-            )
-
+    def _make_gce_instance(self, instance_name):
```
Comment (wallrj) [inline flocker/provision/_gce.py]: :question: Maybe add a docstring here explaining that this function generates a flocker.provisioning shaped record of the GCEInstance from resource queries to the GCE API.

File: flocker/provision/_gce.py [hunk 3]
```diff
@@ -556,8 +609,28 @@ def create_nodes(self, reactor, names, distribution, metadata={}):
         wait_for_operation that uses loop_until instead of poll_until
         to perform the waiting.
```
Comment (wallrj) [inline flocker/provision/_gce.py]: The comment above needs updating.

**Comment by wallrj** [comment]
Thanks @sarum90 

I didn't test the code, just read it. 
I know that you've tested this heavily over the last week and that's good enough for me.
I made a few comments above which you can address or defer as you like.

Please merge when you're ready.

File: flocker/node/agents/gce.py (hunk 1 above)
Comment (sarum90) [inline flocker/node/agents/gce.py]: Done.

File: flocker/provision/_gce.py (hunk 2 above)
Comment (sarum90) [inline flocker/provision/_gce.py]: Added.

File: flocker/provision/_gce.py (hunk 3 above)
Comment (sarum90) [inline flocker/provision/_gce.py]: Removed docstring as it is described in the interface.

**Comment by sarum90** [comment]
Thanks Richard!

I decided to fix up all of the comments, I'll be running merge-pr momentarily.

It'll be nice to have the docs up-to-date. Then I can get rid of that other doc I have laying around in that other repo.

**Comment by sarum90** [comment]
Apologies about the last commit. Looks like GCE changed some of their error codes, the commit is needed for our functional tests to pass.

**Comment by sarum90** [comment]
I've got merge-pr running on this branch. It will correctly wait until we've had passing acceptance tests before merging.

Might be a bit as jenkins likes to kill the slaves the acceptance tests are running on though.

**Pull request merged**